google-cloud-storage/src/main/java/com/google/cloud/storage/Storage.java

```diff
@@ -1255,6 +1260,447 @@ public static SignUrlOption withQueryParams(Map<String, String> queryParams) {
     }
   }
 
+  class V4PostPolicyOption implements Serializable {
+    private static final long serialVersionUID = 8150867146534084543L;
+    private final V4PostPolicyOption.Option option;
+    private final Object value;
+
+    enum Option {
+      PATH_STYLE,
+      VIRTUAL_HOSTED_STYLE,
+      BUCKET_BOUND_HOST_NAME,
+      SERVICE_ACCOUNT_CRED
+    }
+
+    private V4PostPolicyOption(Option option, Object value) {
+      this.option = option;
+      this.value = value;
+    }
+
+    V4PostPolicyOption.Option getOption() {
+      return option;
+    }
+
+    Object getValue() {
+      return value;
+    }
+
+    /**
+     * Provides a service account signer to sign the policy. If not provided an attempt will be made
+     * to get it from the environment.
+     *
+     * @see <a href="https://cloud.google.com/storage/docs/authentication#service_accounts">Service
+     *     Accounts</a>
+     */
+    public static V4PostPolicyOption signWith(ServiceAccountSigner signer) {
+      return new V4PostPolicyOption(V4PostPolicyOption.Option.SERVICE_ACCOUNT_CRED, signer);
+    }
+
+    /**
+     * Use a virtual hosted-style hostname, which adds the bucket into the host portion of the URI
+     * rather than the path, e.g. 'https://mybucket.storage.googleapis.com/...'. The bucket name
+     * will be obtained from the resource passed in.
+     *
+     * @see <a href="https://cloud.google.com/storage/docs/request-endpoints">Request Endpoints</a>
+     */
+    public static V4PostPolicyOption withVirtualHostedStyle() {
+      return new V4PostPolicyOption(V4PostPolicyOption.Option.VIRTUAL_HOSTED_STYLE, "");
+    }
+
+    /**
+     * Generate a path-style URL, which places the bucket name in the path portion of the URL
+     * instead of in the hostname, e.g 'https://storage.googleapis.com/mybucket/...'. Note that this
+     * cannot be used alongside {@code withVirtualHostedStyle()}. Virtual hosted-style URLs, which
+     * can be used via the {@code withVirtualHostedStyle()} method, should generally be preferred
+     * instead of path-style URLs.
+     *
+     * @see <a href="https://cloud.google.com/storage/docs/request-endpoints">Request Endpoints</a>
+     */
+    public static V4PostPolicyOption withPathStyle() {
+      return new V4PostPolicyOption(V4PostPolicyOption.Option.PATH_STYLE, "");
+    }
+
+    /**
+     * Use a bucket-bound hostname, which replaces the storage.googleapis.com host with the name of
+     * a CNAME bucket, e.g. a bucket named 'gcs-subdomain.my.domain.tld', or a Google Cloud Load
+     * Balancer which routes to a bucket you own, e.g. 'my-load-balancer-domain.tld'. Note that this
```

Comment:
No Latin; e.g. --> for example
https://developers.google.com/style/abbreviations